Default serialisation fails for `BatchNorm`.

Hi,

The defaut serialisation fails when a model with `BatchNorm` is serialised. A small test script executed on `dev` branch.

```python

def test_serialise_bn(getkey):
    net = eqx.nn.Sequential(
        [
            eqx.experimental.BatchNorm(3, axis_name="batch"),
        ]
    )

    eqx.tree_serialise_leaves('/tmp/net.eqx', net)

    assert True
```
 
with the error
```console
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../equinox/serialisation.py:183: in tree_serialise_leaves
    jtu.tree_map(_serialise, filter_spec, pytree)
../../../miniconda3/envs/equinox/lib/python3.8/site-packages/jax/_src/tree_util.py:201: in tree_map
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
../../../miniconda3/envs/equinox/lib/python3.8/site-packages/jax/_src/tree_util.py:201: in <genexpr>
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
../equinox/serialisation.py:181: in _serialise
    jtu.tree_map(__serialise, x, is_leaf=is_leaf)
../../../miniconda3/envs/equinox/lib/python3.8/site-packages/jax/_src/tree_util.py:201: in tree_map
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
../../../miniconda3/envs/equinox/lib/python3.8/site-packages/jax/_src/tree_util.py:201: in <genexpr>
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
../equinox/serialisation.py:179: in __serialise
    spec(f, y)
../equinox/serialisation.py:50: in default_serialise_filter_spec
    value, _, _ = x.unsafe_get()
../equinox/experimental/stateful.py:112: in unsafe_get
    return _state_cache[self._obj]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <WeakKeyDictionary at 0x7fc6deff8d90>
key = <equinox.experimental.stateful._IndexObj object at 0x7fc6deec12b0>

    def __getitem__(self, key):
>       return self.data[ref(key)]
E       KeyError: <weakref at 0x7fc736f7e590; to '_IndexObj' at 0x7fc6deec12b0>

../../../miniconda3/envs/equinox/lib/python3.8/weakref.py:383: KeyError
```

--- a/equinox/serialisation.py
+++ b/equinox/serialisation.py
@@ -47,8 +47,13 @@ def default_serialise_filter_spec(f: BinaryIO, x: Any) -> None:
     elif isinstance(x, (bool, float, complex, int)):
         np.save(f, x)
     elif isinstance(x, experimental.StateIndex):
-        value, _, _ = x.unsafe_get()
-        jnp.save(f, value)
+        try:
+            value, _, _ = x.unsafe_get()
+        except KeyError:
+            np.save(f, False)
+        else:
+            np.save(f, True)
+            jnp.save(f, value)
     else:
         pass
 
@@ -91,8 +96,10 @@ def default_deserialise_filter_spec(f: BinaryIO, x: Any) -> Any:
     elif isinstance(x, (bool, float, complex, int)):
         return np.load(f).item()
     elif isinstance(x, experimental.StateIndex):
-        value = jnp.load(f)
-        experimental.set_state(x, value)
+        saved_value = np.load(f)
+        if saved_value:
+            value = jnp.load(f)
+            experimental.set_state(x, value)
         return x
     else:
         return x

--- a/tests/test_serialisation.py
+++ b/tests/test_serialisation.py
@@ -34,7 +34,6 @@ def test_leaf_serialisation(getkey, tmp_path):
     like_numpy_array2 = np.array([6.0, 7.0])
     like_scalars = (False, 6, 6.0, 6 + 6j)
     like_index = eqx.experimental.StateIndex()
-    eqx.experimental.set_state(index, jnp.array(6))
     like_func = lambda x: x
     like_obj = object()
     like = (
@@ -132,3 +131,8 @@ def test_custom_leaf_serialisation(getkey, tmp_path):
     assert tree_loaded_func is unlike_func
     assert tree_loaded_obj is like_obj
     assert tree_loaded_obj is not tree_ser_obj
+
+
+def test_serialise_empty_state(getkey, tmp_path):
+    index = eqx.experimental.StateIndex()
+    eqx.tree_serialise_leaves(tmp_path, index)

